### PR TITLE
[1377] Migrating to new Azure DevOps Extension SDK

### PR DESCRIPTION
**Pull request by phongcao**
Replacing `vss-web-extension-sdk` with `azure-devops-extension-api` and `azure-devops-extension-sdk`.

**Porting Notes**

- Most data contracts stay the same so we just need to change the importing paths.
- Changing the way we invoke services (required by the new sdk) and adding `servicesHelper` file containing helper methods.
- Re-writing logic where _async_ calls are required instead of synchronous ones.
- Most REST client classes have been changed in the new sdk.
- Did manual testing for some basic functions and everything seems to be working.

### CODE REVIEW

File: RetrospectiveExtension.Frontend/dal/azureDevOpsWorkItemService.tsx [hunk 1]
```diff
@@ -85,7 +90,8 @@ class WorkItemService {
       },
     ];
 
-    return this._httpClient.createWorkItem(operation, VSS.getWebContext().project.id, WorkItemService.retrospective_type);
+    return getProjectId()
+      .then(projectId => this._httpClient.createWorkItem(operation, projectId, WorkItemService.retrospective_type));
```
Comment (mhassaninmsft) [inline RetrospectiveExtension.Frontend/dal/azureDevOpsWorkItemService.tsx]: curious why you are using `then` here instead of making the method async and awaiting. any limitations here?

File: RetrospectiveExtension.Frontend/dal/processService.tsx [hunk 2]
```diff
@@ -77,13 +80,13 @@ export class ProcessService {
...
 
     public getProcessForCurrentProject = () => {
-        const currentProjectId = VSS.getWebContext().project.id;
-
-        return this.getProcesses().then(processModels => {
-            return processModels.find((process) => {
-                const projects = process.projects ? process.projects : [];
-                return projects.some((project) => project.id === currentProjectId);
-            });       
+        return getProjectId().then(currentProjectId => {
```
Comment (mhassaninmsft) [inline RetrospectiveExtension.Frontend/dal/processService.tsx:83]: method is great candidate for async/ await refactor if the context allows it

File: RetrospectiveExtension.Frontend/utilities/external/telemetryclient.ts [hunk 3]
```diff
@@ -52,10 +52,10 @@ export class TelemetryClient {
     this.ExtensionContext = settings.extensioncontext;
 
     try {
-      const webContext = VSS.getWebContext();
+      const user = SDK.getUser();
 
       AppInsights.downloadAndSetup(config);
-      AppInsights.setAuthenticatedUserContext(webContext.user.id, webContext.collection.id);
+      AppInsights.setAuthenticatedUserContext(user.id, user.name);
```
Comment (mhassaninmsft) [inline RetrospectiveExtension.Frontend/utilities/external/telemetryclient.ts:58]: what is the difference between `collection.id` and `user.name`? is this an intended change?

File: RetrospectiveExtension.Frontend/vss-extension-dev.json [hunk 4]
```diff
@@ -72,10 +72,6 @@
       "path": "src",
       "addressable": true
     },
-    {
```
Comment (mhassaninmsft) [inline RetrospectiveExtension.Frontend/vss-extension-dev.json:75]: curious why the `azure-devops-extension` library was not included here as well

File: RetrospectiveExtension.Frontend/webpack.config.js [hunk 5]
```diff
@@ -12,11 +12,7 @@ module.exports = {
   output: {
     path: BUILD_DIR,
     filename: './reflect-bundle.js',
-    libraryTarget: 'amd'
```
Comment (mhassaninmsft) [inline RetrospectiveExtension.Frontend/webpack.config.js:17]: why was this removed?

**Review by mhassaninmsft — COMMENTED** [review]
Looks great Phong!

File: RetrospectiveExtension.Frontend/vss-extension-dev.json (hunk 4 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/vss-extension-dev.json:75]: I'm not sure about the old vss sdk and why it's there from the beginning. However, for the new one, there are 2 reasons:

- When we use webpack to create the bundle file, it's already contained all info it needs from those sdks/apis.
- The sdk itself is just a shim to invoke the underlying services which are only available in an Azure DevOps environment so I guess that there is no need for them in runtime. Also, the sample codes available don't include them.

File: RetrospectiveExtension.Frontend/webpack.config.js (hunk 5 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/webpack.config.js:17]: If you include it, the bundle js file will be built and expected to be loaded in a requirejs environment. With the old initialization, the `vss.load(..)` method can handle the bundle file but in the new sdk, it'll cause the `define is not defined` error. You can have a look at the `index.html` and `index.tsx` files for the initialization process.

File: RetrospectiveExtension.Frontend/utilities/external/telemetryclient.ts (hunk 3 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/utilities/external/telemetryclient.ts:58]: Yes, they are. One is the unique id and the other is email address.

File: RetrospectiveExtension.Frontend/dal/processService.tsx (hunk 2 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/dal/processService.tsx:83]: The old code actually uses `.then` so when I changed it, I want to follow the same pattern. If we want to switch to async, we should let @polatengin decide that and do it for the whole file for consistency.

File: RetrospectiveExtension.Frontend/dal/azureDevOpsWorkItemService.tsx (hunk 1 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/dal/azureDevOpsWorkItemService.tsx]: This file actually contains both `async` and `.then` styles. @polatengin what's your thoughts? If we're going to change to async, it'll make the code more readable so something to consider for future PR?

File: RetrospectiveExtension.Frontend/components/actionItem.tsx [hunk 6]
```diff
@@ -8,8 +8,9 @@ import {
   IDocumentCardPreviewProps
 } from 'office-ui-fabric-react/lib/DocumentCard';
 import * as React from 'react';
-import { WorkItem, WorkItemType, WorkItemStateColor } from 'TFS/WorkItemTracking/Contracts';
-import { WorkItemFormNavigationService } from 'TFS/WorkItemTracking/Services';
+import * as SDK from 'azure-devops-extension-sdk';
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/components/actionItem.tsx]: do we need everything from SDK? if we can capture only the methods we're using, tree-shaker and code-trimmers might spit fewer bytes during the compile process.

File: RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx [hunk 7]
```diff
@@ -181,6 +182,9 @@ export default class FeedbackBoardContainer extends React.Component<FeedbackBoar
         setTimeout(this.tryReconnectToBackend, 2000);
       });
 
+      const isHosted = await isHostedAzureDevOps();
+      this.setState({ isHostedAzureDevOps: isHosted });
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx]: instead of setting default value for the `isHostedAzureDevOps` state and calculate it and setting the state afterwards, can we do the calculation beforehand and use it when we set the default state values?

something like (in pseudo);

- calculate if it's hosted on azure devops
- set the default state

it could save a re-render due to the extra state update

File: RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx [hunk 8]
```diff
@@ -718,8 +722,8 @@ export default class FeedbackBoardContainer extends React.Component<FeedbackBoar
   }
 
   private createBoard = async (title: string, maxvotesPerUser: number, columns: IFeedbackColumn[], isIncludeTeamEffectivenessMeasurement: boolean, isBoardAnonymous: boolean, shouldShowFeedbackAfterCollect: boolean, displayPrimeDirective: boolean) => {
-const createdBoard = await BoardDataService.createBoardForTeam(this.state.currentTeam.id, title, maxvotesPerUser, columns, isIncludeTeamEffectivenessMeasurement, isBoardAnonymous, shouldShowFeedbackAfterCollect, displayPrimeDirective);
-console.log(createdBoard);
+    const createdBoard = await BoardDataService.createBoardForTeam(this.state.currentTeam.id, title, maxvotesPerUser, columns, isIncludeTeamEffectivenessMeasurement, isBoardAnonymous, shouldShowFeedbackAfterCollect, displayPrimeDirective);
+    console.log(createdBoard);
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx]: did we forget to remove debug related `console.log`? if it's for debugging purposes, I'd like having no `console.log` in the code, if it's not for debugging, we should a sentence that describes what's the log for.

File: RetrospectiveExtension.Frontend/dal/azureDevOpsWorkItemService.tsx (hunk 1 above)
Comment (polatengin) [reply RetrospectiveExtension.Frontend/dal/azureDevOpsWorkItemService.tsx]: I'd like to have a single approach, most of the codebase is using `async-await` pattern.

if we can make the method `async`, we should definitely convert this to `await`, instead of using `.then()` functions.

sometimes, we can't convert method to `async` method, in those cases, we can use `.then()` functions.

File: RetrospectiveExtension.Frontend/dal/azureDevOpsWorkService.tsx [hunk 9]
```diff
@@ -47,9 +51,10 @@ class WorkService {
    */
   public async getTeamFieldValues(teamId: string):
     Promise<TeamFieldValues> {
+    const projectId = await getProjectId();
     const teamContext = {
       project: '',
-      projectId: VSS.getWebContext().project.id,
+      projectId: projectId,
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/dal/azureDevOpsWorkService.tsx]: on line 25 it's only `projectId`, here it's `projectId: projectId`

both of them work, but I think we should use the same style. I suggest to pick one of the styles and use it.

File: RetrospectiveExtension.Frontend/dal/processService.tsx [hunk 10]
```diff
@@ -77,13 +80,13 @@ export class ProcessService {
     }
 
     public getProcessForCurrentProject = () => {
```
Comment (polatengin) [inline RetrospectiveExtension.Frontend/dal/processService.tsx:82]: if the caller can `await`, this method also can be turned to an `async` method.

when I started coding this extension, I wasn't super familiar with `async-await`, I might missing some obvious awaitable methods.

**Review by polatengin — CHANGES_REQUESTED** [review]
it looks GREAT, although I left a few comments. if we can address them, that would be perfect! 👍

File: RetrospectiveExtension.Frontend/components/actionItem.tsx (hunk 6 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/components/actionItem.tsx]: Let me check and fix other places as well.

File: RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx (hunk 7 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx]: I'll move it to index.tsx file.

File: RetrospectiveExtension.Frontend/dal/azureDevOpsWorkService.tsx (hunk 9 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/dal/azureDevOpsWorkService.tsx]: Good catch! Fixed.

File: RetrospectiveExtension.Frontend/dal/azureDevOpsWorkItemService.tsx (hunk 1 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/dal/azureDevOpsWorkItemService.tsx]: Converted several methods that use `getProjectId` to _async_ type. For the rest of the file, we should fix them in a separate PR since it's out of scope for this one.

File: RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx (hunk 8 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx]: It's actually there in the original code. Happy to remove it now that you mention it.

File: RetrospectiveExtension.Frontend/dal/processService.tsx (hunk 10 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/dal/processService.tsx:82]: I suggest we keep this change as it's consistent with the rest of the file (all of them uses `.then` style). If we want to switch to async, we can do that in a separate PR in the future.

File: RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx (hunk 7 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx]: Fixing it by moving `isHostedAzureDevOps` to props from `index.tsx` file.

File: RetrospectiveExtension.Frontend/components/actionItem.tsx (hunk 6 above)
Comment (phongcao) [reply RetrospectiveExtension.Frontend/components/actionItem.tsx]: Seems to me that webpack is smart enough to only include what are needed so the bundle size doesn't change after I've fixed all of them. However, it's still a good practice to only include what you use in the file.

**Review by polatengin — APPROVED** [review]
looks perfect, thanks 🚀💯

**Comment by phongcao** [comment]
@mhassaninmsft Can you have a look again?

**Review by mhassaninmsft — APPROVED** [review]
